api/server/v1/error.go

```diff
@@ -17,98 +17,292 @@ package api
+//        the GRPC status (see MarshalStatus) in GRPC gateway
+//      * Client parses the error using UnmarshalStatus into TigrisError
+
 // TigrisError is our Tigris HTTP counterpart of grpc status. All the APIs should use this Error to return as a user facing
 // error. TigrisError will return grpcStatus to the muxer so that grpc client can see grpcStatus as the final output. For
 // HTTP clients see the **MarshalStatus** method where we are returning the response by not the grpc code as that is not
 // needed for HTTP clients.
 type TigrisError struct {
 	// The status code, which should be an enum value of [google.rpc.Code][google.rpc.Code]. We don't need to marshal
 	// this code for HTTP clients.
-	Code codes.Code `json:"code,omitempty"`
+	Code Code `json:"code,omitempty"`
 	// A developer-facing error message.
 	Message string `json:"message,omitempty"`
-	// A list of messages that carry the error details. This is mainly to send our internal codes and messages.
-	Details []*ErrorDetails `json:"details,omitempty"`
+
+	Details []proto.Message `json:"details,omitempty"`
 }
 
 // Error to return the underlying error message
 func (e *TigrisError) Error() string {
 	return e.Message
 }
 
-// GRPCStatus converts the TigrisError and return status.Status. This is used to return grpc status to the grpc clients
-func (e *TigrisError) GRPCStatus() *status.Status {
-	s := &spb.Status{
-		Code:    int32(e.Code),
-		Message: e.Message,
-	}
+// WithDetails a helper method for adding details to the TigrisError
+func (e *TigrisError) WithDetails(details ...proto.Message) *TigrisError {
+	e.Details = append(e.Details, details...)
+	return e
+}
 
-	if len(e.Details) == 0 {
-		return status.FromProto(s)
+// WithRetry attached retry information to the error
+func (e *TigrisError) WithRetry(d time.Duration) *TigrisError {
+	if d != 0 {
+		e.Details = append(e.Details, &errdetails.RetryInfo{RetryDelay: durationpb.New(d)})
 	}
+	return e
+}
+
+// RetryDelay retrieves retry delay if it's attached to the error
+func (e *TigrisError) RetryDelay() time.Duration {
+	var dur time.Duration
 
-	var details []*any.Any
 	for _, d := range e.Details {
-		var a = &any.Any{}
-		err := anypb.MarshalFrom(a, d, proto.MarshalOptions{})
-		if err == nil {
-			details = append(details, a)
+		switch t := d.(type) {
+		case *errdetails.RetryInfo:
+			dur = t.RetryDelay.AsDuration()
 		}
 	}
 
-	s.Details = details
-	return status.FromProto(s)
+	return dur
 }
 
-// WithDetails a helper method for adding details to the TigrisError
-func (e *TigrisError) WithDetails(details ...*ErrorDetails) *TigrisError {
-	e.Details = details
-	return e
+// ToGRPCCode converts Tigris error code to GRPC code
+// Extended codes converted to 'Unknown' GRPC code
+func ToGRPCCode(code Code) codes.Code {
+	if code <= Code_UNAUTHENTICATED {
+		return codes.Code(code)
+	}
+
+	// Cannot be represented by GRPC codes
+	return codes.Unknown
```

Comment:
So how will gRPC client know the actual code in this case?

Reply:
It will be reconstructed on the client from errdetails.ErrorInfo, see FromStatusError error below.